### ISSUE TITLE
Investigate why older shell releases are failing against SNAPSHOT server

### ISSUE THREAD

**Issue by sabbyanandan** [spring-cloud/spring-cloud-dataflow#1129]
As a user, I'm trying to connect to the server with an incompatible version of the shell and I see the following error. The error message can be improved with pointers on where to get the compatible shell bits. 

**Error:**
```
" Unable to contact Data Flow Server at 'https://dataflow-server-kl.cfapps.io': 'java.lang.NullPointerException'."
```

**Acceptance:**
- A user-friendly error message stating _"You're trying to connect to the server with an incompatible shell - please download the compatible version from http://link ..... "_
- The download link points to the compatible shell release

**Comment by jvalkeal** [comment]
I was thinking that It's probably not a good idea as is because we don't know if client/server is incompatible or if there's a bug.

Maybe we should finally version our rest api and then come up with a versioning scheme so that server could deny 'wrong' clients. It's kinda fundamental thing that either you support one api version or keep multiple api versions alive for older clients.

Maybe we should change a scope of this ticket?

**Comment by sabbyanandan** [comment]
@jvalkeal: The shell is a sub-module in core, so whenever we release SCDF-core, it is implicit that we will have an equivalent shell release. I was hoping to make a simple improvement to resolve the shell version automatically and add a download link in shell/ui. 

There's a lot of chatter due to incompatible shell + server combo. We could make it simpler for the users to download the right shell release. Perhaps when we will have a lot more APIs to support, we could revisit the versioning angle. Thoughts?

**Comment by sabbyanandan** [comment]
See this WIP from @ghillert on the UI -> https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/150

**Comment by ghillert** [comment]
Just added the `bug` label ... `NullPointerException` always qualifies as bug. 

At this stage, I would be hesitant wanting to support versioning the the REST API itself (and e.g. version ranges of compatible client/server combos) as the API is still evolving and also considering the increased maintenance burden.

Nonetheless, the current user-experience is bad. We should at least make the target command send the version info of the shell to the server during `target` and deny the client request with a proper message if the versions don’t match. Maybe even include a `flag` to bi-pass that for special use-cases (e.g. development).

The shell is a pretty trivial download. We could even include the download link in the shell error message to improve the user-experience.

**Comment by trisberg** [comment]
Do we know what versions throw this NPE? I'm not able to reproduce - I do get this:

```
server-unknown:>dataflow config server --uri http://localhost:9393
Unable to contact Data Flow Server at 'http://localhost:9393': 'java.lang.IllegalArgumentException: Deployments relation is required'.
```

This is using Shell 1.1.2 and latest build snapshot for 1.2.0 (1.2.0.BUILD-20170206.094104-13)

**Comment by trisberg** [comment]
Looks like the TaskTemplate fails on creation since we have removed the "tasks/deployments" link in 1.2.x. Going forward we should decide the policy for shell-server compatibility and, like Janne suggests, do some version negotiation on shell URI configuration and display the proper compatibility requirement messages

**Comment by trisberg** [comment]
Created two issues to fix this:  #1191 and #1192
